text/0000-closure-lifetime-binder.md

```diff
@@ -0,0 +1,266 @@
+- Feature Name: `closure_lifetime_binder`
+- Start Date: 2022-01-06
+- RFC PR: [rust-lang/rfcs#0000](https://github.com/rust-lang/rfcs/pull/0000)
+- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)
+
+
+This RFC went through a pre-RFC phase at https://internals.rust-lang.org/t/pre-rfc-allow-for-a-syntax-with-closures-for-explicit-higher-ranked-lifetimes/15888
+
+# Summary
+
+Allow explicitly specifying lifetimes on closures via `for<'a> |arg: &'a u8| { ... }`. This will always result in a higher-ranked closure which can accept *any* lifetime (as in `fn bar<'a>(val: &'a u8) {}`). Closures defined without the `for<'a>` syntax retain their current behavior: lifetimes will be inferred as either some local region (via an inference variable), or a higher-ranked lifetime.
+
+# Motivation
+
+There are several open issues around closure lifetimes (https://github.com/rust-lang/rust/issues/91966 and https://github.com/rust-lang/rust/issues/41078), all of which stem from type inference incorrectly choosing either a higher-ranked lifetime, or a local lifetime.
+
+This can be illustrated in the following cases:
+
+1. We infer a higher-ranked region ( `for<'a> fn(&'a u8)` ) when we really want some specific (local) region. This occurs in the following code:
+
+```rust
+fn main () {
+    let mut fields: Vec<&str> = Vec::new();
+    let pusher = |a: &str| fields.push(a);
+}
+```
+
+which gives the error:
+
+```
+error[E0521]: borrowed data escapes outside of closure
+ --> src/main.rs:3:28
+  |
+2 |     let mut fields: Vec<&str> = Vec::new();
+  |         ---------- `fields` declared here, outside of the closure body
+3 |     let pusher = |a: &str| fields.push(a);
+  |                   -        ^^^^^^^^^^^^^^ `a` escapes the closure body here
+  |                   |
+  |                   `a` is a reference that is only valid in the closure body
+```
+
+The issue is that `Vec<&str>` is not higher-ranked, so we can only push an `&'0 str` for some specific lifetime `'0` . The `pusher` closure signature requires that it accept *any* lifetime, which leads to a compiler error.
+
+2. We infer some specific region when we really want a higher-ranked region. This occurs in the following code:
+
+```rust
+use std::cell::Cell;
+
+fn main() {
+    let static_cell: Cell<&'static u8> = Cell::new(&25);
+    let closure = |s| {};
+    closure(static_cell);
+    let val = 30;
+    let short_cell: Cell<&u8> = Cell::new(&val);
+    closure(short_cell);
+}
+```
```

Comment:
FWIW, there is no need for invariance to show the problems of an inferred / non-higher-order lifetime parameter:

```rust
fn main ()
{
    let closure = |s| {
        let _: &'_ i32 = s; // type-annotations outside the param list don't help.
    };
    {
        let local = 42;
        closure(&local);
    }
    {
        let local = 42;
        closure(&local);
    }
}
```

fails as well.

  - [Playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=29fa3ec50135c22b0077138468c1adc6)

or even shorter:

```rust
let closure = |_| ();
closure(&i32::default());
closure(&i32::default());
```